src/Shape.jl

```diff
@@ -0,0 +1,275 @@
+end
+
+function Shape(num_points::Int, points_of_subpaths::Vector{Int})
+    points = Vector{Point}(undef, num_points)
+    subpaths = Vector{Vector{Point}}()
+    for points_in_subpath in points_of_subpaths
+        push!(subpaths, Vector{Point}(undef, points_in_subpath))
+    end
+    return Shape(points, Point[], O, Point[], subpaths, 0, 0, 0)
+end
+
+function Shape(points, subpaths; compute_info = true)
+    simplified_points = simplify(points)
+    if compute_info
+        points, aa, oa, ra = get_angles(simplified_points)
+    else
+        aa = 0
+        oa = 0
+        ra = 0
+    end
+    centroid = polycentroid(points)
+    shape =
+        Shape(points, simplified_points, centroid, points .- centroid, subpaths, aa, oa, ra)
+    return shape
+end
+
+function Base.isempty(shape::Shape)
+    return length(shape.points) == 0
+end
+
+
+function get_angles(p)
+    # TODO: let's assume it's a closed path
+    num_acute_angles = 0
+    num_obtuse_angles = 0
+    num_right_angles = 0
+    simplified = Point[]
+
+    for i in 1:length(p)
+        pA = p[i]
+        if i < length(p) - 1
+            u = p[i] - p[i + 1]
+            v = p[i + 2] - p[i + 1]
+            pB = p[i + 1]
+            pC = p[i + 2]
+        elseif i == length(p) - 1
+            u = p[i] - p[i + 1]
+            v = p[1] - p[i + 1]
+            pB = p[i + 1]
+            pC = p[1]
+        else
+            u = p[i] - p[1]
+            v = p[2] - p[1]
+            pB = p[1]
+            pC = p[2]
+        end
+        ang = rad2deg(acos(clamp(
+            dotproduct(u, v) / (sqrt(u.x^2 + u.y^2) * sqrt(v.x^2 + v.y^2)),
+            -1,
+            1,
+        )))
+        if ang < 179.8
+            push!(simplified, pB)
+        end
+
+        if ang < 160
+            if ang > 100 # obtuse
+                num_obtuse_angles += 1
+            elseif ang < 80 # acute
+                num_acute_angles += 1
+            else # right
+                num_right_angles += 1
+            end
+        end
+    end
+    return simplified, num_acute_angles, num_obtuse_angles, num_right_angles
+end
+
+function get_similarity(shapeA::Shape, shapeB::Shape)
+    if isempty(shapeA) || isempty(shapeB)
+        return 0.0
+    end
+    score_points = 10.0
+    score_holes = 500.0
+    score_acute_angles = 100.0
+    score_right_angles = 100.0
+    score_obtuse_angles = 100.0
+    score_point_diff = 100.0
+    score_centered_point_diff = 100.0
+
+    score = 0.0
+
+    # number of points
+    nA = length(shapeA.simplified_points)
+    nB = length(shapeB.simplified_points)
+    perc = nA < nB ? nA / nB : nB / nA
+    perc = isnan(perc) ? 1.0 : perc
+    # println("Point score: ", perc*score_points)
+    score += perc * score_points
+
+    # number of holes
+    nA = length(shapeA.subpaths)
+    nB = length(shapeB.subpaths)
+    perc = nA < nB ? nA / nB : nB / nA
+    perc = isnan(perc) ? 1.0 : perc
+    # println("Hole score: ", perc*score_holes)
```

Comment:
I see that there a lot of `println` statements commented out in this code. Could you remove them if they are no longer needed?

Reply:
sure